src/Tools/Source/CompilerGeneratorTools/Source/CSharpSyntaxGenerator/AbstractFileWriter.cs

```diff
@@ -119,14 +119,32 @@ protected static bool CanBeField(Field field)
 
         protected static string GetFieldType(Field field, bool green)
         {
-            if (IsAnyList(field.Type))
+            // Fields in red trees are lazily initialized, with null as the uninitialized value
+            return GetNullableType(field.Type, optionalOrLazy: IsOptional(field) || !green, green);
+        }
+
+        private static string GetNullableType(string fieldType, bool optionalOrLazy, bool green)
+        {
+            if (IsAnyList(fieldType))
             {
-                return green
-                    ? "GreenNode"
-                    : "SyntaxNode";
+                if (optionalOrLazy)
+                    return green ? "GreenNode?" : "SyntaxNode?";
+                else
+                    return green ? "GreenNode?" : "SyntaxNode";
```

Comment:
why `SyntaxNode` here and not `SyntaxNode?`?

Reply:
This is really misleading though, given that the function is titled `GetNullableType`. Consider either renaming the function or changing the calling pattern to make the name hold true.

Reply:
Renamed and made it a local function to avoid the confusion.